spring-data-jpa/src/main/java/org/springframework/data/jpa/repository/query/QueryUtils.java

```diff
@@ -821,20 +822,28 @@ private static boolean requiresOuterJoin(From<?, ?> from, PropertyPath property,
 		Bindable<?> propertyPathModel;
 		Bindable<?> model = from.getModel();
 
-		// required for EclipseLink: we try to avoid using from.get as EclipseLink produces an inner join
-		// regardless of which join operation is specified next
-		// see: https://bugs.eclipse.org/bugs/show_bug.cgi?id=413892
-		// still occurs as of 2.7
-		ManagedType<?> managedType = null;
-		if (model instanceof ManagedType) {
-			managedType = (ManagedType<?>) model;
-		} else if (model instanceof SingularAttribute
-				&& ((SingularAttribute<?, ?>) model).getType() instanceof ManagedType) {
-			managedType = (ManagedType<?>) ((SingularAttribute<?, ?>) model).getType();
-		}
-		if (managedType != null) {
-			propertyPathModel = (Bindable<?>) managedType.getAttribute(segment);
+		if (model.getClass().getName().startsWith("org.eclipse.persistence")) {
+			// required for EclipseLink: we try to avoid using from.get as EclipseLink produces an inner join
+			// regardless of which join operation is specified next
+			// see: https://bugs.eclipse.org/bugs/show_bug.cgi?id=413892
+			// still occurs as of 2.7
+			ManagedType<?> managedType = null;
+			if (model instanceof ManagedType) {
+				managedType = (ManagedType<?>) model;
+			} else if (model instanceof SingularAttribute
+					&& ((SingularAttribute<?, ?>) model).getType() instanceof ManagedType) {
+				managedType = (ManagedType<?>) ((SingularAttribute<?, ?>) model).getType();
+			}
+			if (managedType != null) {
+				propertyPathModel = (Bindable<?>) managedType.getAttribute(segment);
+			} else {
+				propertyPathModel = from.get(segment).getModel();
+			}
 		} else {
+			// ManagedType may be erased type for some vendor if the attribute is declared as generic
+			// see: https://hibernate.atlassian.net/browse/HHH-16144
+			// see: https://github.com/hibernate/hibernate-orm/pull/7630
+			// see: https://github.com/jakartaee/persistence/issues/562
```

Comment:
I'm curious if we could hide the vendor specifics behind eg. the `PersistenceProvider` so we can encapsulate different behaviours when getting hold of the attribute .

I'm thinking of something that would allow us to do something like the following:
```java
Bindable<?> propertyPathModel = PersistenceProvider.segmentResolver(from).resolve(segment);
``` 
Having `PersistenceProvider` implement a let's call it `PathSegmentResolver` defaulting to `from.get(segment).getModel()`, we could move the eclipselink specific requirements with all the type checks to the ECLIPSELINK implementation.

Reply:
How to resolve `PersistenceProvider` in method `QueryUtils::requiresOuterJoin`?